Subject: [PATCH] dhcp: additional options, ported for busybox 1.36.1

Enable send/receive of additional DHCP options (Curt Brune, ONIE):
DHCP_LOG_SERVER (0x07), DHCP_VENDOR_OPTS (0x2b), DHCP_WWW_SERVER
(0x48), DHCP_DEFAULT_URL (0x72), DHCP_VIVSO_OPTS (0x7d),
DHCP_TFTP_SERVER_IP (0x96).

Rebased onto the 1.36.1 tables (upstream added PCODE/TCODE/REBOOT_TIME
entries in between; OPTION_BIN type already exists upstream).
diff --git a/networking/udhcp/common.c b/networking/udhcp/common.c
index ae818db..b83390c 100644
--- a/networking/udhcp/common.c
+++ b/networking/udhcp/common.c
@@ -27,7 +27,7 @@ const struct dhcp_optflag dhcp_optflags[] ALIGN2 = {
 //	{ OPTION_IP | OPTION_LIST                 , 0x04 }, /* DHCP_TIME_SERVER   */
 //	{ OPTION_IP | OPTION_LIST                 , 0x05 }, /* DHCP_NAME_SERVER   */
 	{ OPTION_IP | OPTION_LIST     | OPTION_REQ, 0x06 }, /* DHCP_DNS_SERVER    */
-//	{ OPTION_IP | OPTION_LIST                 , 0x07 }, /* DHCP_LOG_SERVER    */
+	{ OPTION_IP | OPTION_LIST                 , 0x07 }, /* DHCP_LOG_SERVER    */
 //	{ OPTION_IP | OPTION_LIST                 , 0x08 }, /* DHCP_COOKIE_SERVER */
 	{ OPTION_IP | OPTION_LIST                 , 0x09 }, /* DHCP_LPR_SERVER    */
 	{ OPTION_STRING_HOST          | OPTION_REQ, 0x0c }, /* DHCP_HOST_NAME     */
@@ -45,7 +45,8 @@ const struct dhcp_optflag dhcp_optflags[] ALIGN2 = {
 	{ OPTION_STRING_HOST                      , 0x28 }, /* DHCP_NIS_DOMAIN    */
 	{ OPTION_IP | OPTION_LIST                 , 0x29 }, /* DHCP_NIS_SERVER    */
 	{ OPTION_IP | OPTION_LIST     | OPTION_REQ, 0x2a }, /* DHCP_NTP_SERVER    */
-	{ OPTION_IP | OPTION_LIST                 , 0x2c }, /* DHCP_WINS_SERVER   */
+	{ OPTION_BIN                              , 0x2b }, /* DHCP_VENDOR_OPTS   */
+	{ OPTION_IP | OPTION_LIST                 , 0x2c }, /* DHCP_WINS_SERVER    */
 	{ OPTION_U32                              , 0x33 }, /* DHCP_LEASE_TIME    */
 	{ OPTION_IP                               , 0x36 }, /* DHCP_SERVER_ID     */
 	{ OPTION_STRING                           , 0x38 }, /* DHCP_ERR_MESSAGE   */
@@ -53,8 +54,10 @@ const struct dhcp_optflag dhcp_optflags[] ALIGN2 = {
 //TODO: must be combined with 'sname' and 'file' handling:
 	{ OPTION_STRING_HOST                      , 0x42 }, /* DHCP_TFTP_SERVER_NAME */
 	{ OPTION_STRING                           , 0x43 }, /* DHCP_BOOT_FILE     */
+	{ OPTION_IP | OPTION_LIST                 , 0x48 }, /* DHCP_WWW_SERVER    */
 //TODO: not a string, but a set of LASCII strings:
 //	{ OPTION_STRING                           , 0x4D }, /* DHCP_USER_CLASS    */
+	{ OPTION_STRING                           , 0x72 }, /* DHCP_DEFAULT_URL   */
 	{ OPTION_STRING                           , 0x64 }, /* DHCP_PCODE         */
 	{ OPTION_STRING                           , 0x65 }, /* DHCP_TCODE         */
 #if ENABLE_FEATURE_UDHCP_RFC3397
@@ -62,6 +65,7 @@ const struct dhcp_optflag dhcp_optflags[] ALIGN2 = {
 	{ OPTION_SIP_SERVERS                      , 0x78 }, /* DHCP_SIP_SERVERS   */
 #endif
 	{ OPTION_STATIC_ROUTES | OPTION_LIST      , 0x79 }, /* DHCP_STATIC_ROUTES */
+	{ OPTION_BIN                              , 0x7d }, /* DHCP_VIVSO_OPTS    */
 #if ENABLE_FEATURE_UDHCP_8021Q
 	{ OPTION_U16                              , 0x84 }, /* DHCP_VLAN_ID       */
 	{ OPTION_U8                               , 0x85 }, /* DHCP_VLAN_PRIORITY */
@@ -70,6 +74,7 @@ const struct dhcp_optflag dhcp_optflags[] ALIGN2 = {
 	{ OPTION_STRING                           , 0xd2 }, /* DHCP_PXE_PATH_PREFIX */
 	{ OPTION_U32                              , 0xd3 }, /* DHCP_REBOOT_TIME   */
 	{ OPTION_6RD                              , 0xd4 }, /* DHCP_6RD           */
+	{ OPTION_IP                               , 0x96 }, /* DHCP_TFTP_SERVER_IP */
 	{ OPTION_STATIC_ROUTES | OPTION_LIST      , 0xf9 }, /* DHCP_MS_STATIC_ROUTES */
 	{ OPTION_STRING                           , 0xfc }, /* DHCP_WPAD          */
 
@@ -101,7 +106,7 @@ const char dhcp_option_strings[] ALIGN1 =
 //	"timesrv" "\0"          /* DHCP_TIME_SERVER     */
 //	"namesrv" "\0"          /* DHCP_NAME_SERVER     */
 	"dns" "\0"              /* DHCP_DNS_SERVER      */
-//	"logsrv" "\0"           /* DHCP_LOG_SERVER      */
+	"logsrv" "\0"           /* DHCP_LOG_SERVER      */
 //	"cookiesrv" "\0"        /* DHCP_COOKIE_SERVER   */
 	"lprsrv" "\0"           /* DHCP_LPR_SERVER      */
 	"hostname" "\0"         /* DHCP_HOST_NAME       */
@@ -116,6 +121,7 @@ const char dhcp_option_strings[] ALIGN1 =
 	"nisdomain" "\0"        /* DHCP_NIS_DOMAIN      */
 	"nissrv" "\0"           /* DHCP_NIS_SERVER      */
 	"ntpsrv" "\0"           /* DHCP_NTP_SERVER      */
+	"vendoropts" "\0"       /* DHCP_VENDOR_OPTS     */
 	"wins" "\0"             /* DHCP_WINS_SERVER     */
 	"lease" "\0"            /* DHCP_LEASE_TIME      */
 	"serverid" "\0"         /* DHCP_SERVER_ID       */
@@ -123,7 +129,9 @@ const char dhcp_option_strings[] ALIGN1 =
 	"vendor" "\0"           /* DHCP_VENDOR          */
 	"tftp" "\0"             /* DHCP_TFTP_SERVER_NAME*/
 	"bootfile" "\0"         /* DHCP_BOOT_FILE       */
+	"wwwsrv" "\0"           /* DHCP_WWW_SERVER      */
 //	"userclass" "\0"        /* DHCP_USER_CLASS      */
+	"url" "\0"              /* DHCP_DEFAULT_URL     */
 	"tzstr" "\0"            /* DHCP_PCODE           */
 	"tzdbstr" "\0"          /* DHCP_TCODE           */
 #if ENABLE_FEATURE_UDHCP_RFC3397
@@ -133,6 +141,7 @@ const char dhcp_option_strings[] ALIGN1 =
 	"sipsrv" "\0"           /* DHCP_SIP_SERVERS     */
 #endif
 	"staticroutes" "\0"     /* DHCP_STATIC_ROUTES   */
+	"vivso" "\0"            /* DHCP_VIVSO_OPTS      */
 #if ENABLE_FEATURE_UDHCP_8021Q
 	"vlanid" "\0"           /* DHCP_VLAN_ID         */
 	"vlanpriority" "\0"     /* DHCP_VLAN_PRIORITY   */
@@ -141,6 +150,7 @@ const char dhcp_option_strings[] ALIGN1 =
 	"pxepathprefix" "\0"    /* DHCP_PXE_PATH_PREFIX */
 	"reboottime" "\0"       /* DHCP_REBOOT_TIME     */
 	"ip6rd" "\0"            /* DHCP_6RD             */
+	"tftpsiaddr" "\0"       /* DHCP_TFTP_SERVER_IP  */
 	"msstaticroutes" "\0"   /* DHCP_MS_STATIC_ROUTES*/
 	"wpad" "\0"             /* DHCP_WPAD            */
 	;
@@ -158,6 +168,7 @@ const uint8_t dhcp_option_lengths[] ALIGN1 = {
 	[OPTION_IP] =      4,
 	[OPTION_IP_PAIR] = 8,
 //	[OPTION_BOOLEAN] = 1,
+	[OPTION_BIN] =     1,  /* ignored by udhcp_str2optset */
 	[OPTION_STRING] =  1,  /* ignored by udhcp_str2optset */
 	[OPTION_STRING_HOST] = 1,  /* ignored by udhcp_str2optset */
 #if ENABLE_FEATURE_UDHCP_RFC3397
diff --git a/networking/udhcp/dhcpc.c b/networking/udhcp/dhcpc.c
index c757fb3..e0309d2 100644
--- a/networking/udhcp/dhcpc.c
+++ b/networking/udhcp/dhcpc.c
@@ -126,6 +126,7 @@ static const uint8_t len_of_option_as_string[] ALIGN1 = {
 	[OPTION_IP_PAIR         ] = sizeof("255.255.255.255 ") * 2,
 	[OPTION_STATIC_ROUTES   ] = sizeof("255.255.255.255/32 255.255.255.255 "),
 	[OPTION_6RD             ] = sizeof("132 128 ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff 255.255.255.255 "),
+	[OPTION_BIN             ] = 2,
 	[OPTION_STRING          ] = 1,
 	[OPTION_STRING_HOST     ] = 1,
 #if ENABLE_FEATURE_UDHCP_RFC3397
